### ISSUE TITLE
MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.

### ISSUE THREAD

**Issue by florence26** [lvwerra/trl#440]
I have been trying to run the example code below:

```
# imports
import torch
from transformers import AutoTokenizer
from trl import PPOTrainer, PPOConfig, AutoModelForCausalLMWithValueHead, create_reference_model
from trl.core import respond_to_batch

# get models
model = AutoModelForCausalLMWithValueHead.from_pretrained('gpt2')
model_ref = create_reference_model(model)

tokenizer = AutoTokenizer.from_pretrained('gpt2')
tokenizer.pad_token = tokenizer.eos_token

# initialize trainer
ppo_config = PPOConfig(
    batch_size=1,
)

# encode a query
query_txt = "This morning I went to the "
query_tensor = tokenizer.encode(query_txt, return_tensors="pt")

# get model response
response_tensor  = respond_to_batch(model, query_tensor)

# create a ppo trainer
ppo_trainer = PPOTrainer(ppo_config, model, model_ref, tokenizer)

# define a reward for response
# (this could be any reward such as human feedback or output from another model)
reward = [torch.tensor(1.0)]

# train model for one step with ppo
train_stats = ppo_trainer.step([query_tensor[0]], [response_tensor[0]], reward)
```

I am getting a lot of issues with running Mac’s MPS backend to run this code. I have tried to set `device = ‘cpu’` instead and added
.to(device) or set device = device to the model, tokenizer, query tensor, response tensor and reward, but am getting the error below:

`MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

If I’m not using the MPS Tensor, why is this being shown?

Everything works until the last ppo_trainer.step step.

**Comment by younesbelkada** [comment]
Hi @florence26 
Thank you very much for your issue.
The PPOTrainer calls `accelerator.prepare(model)` under the hood, I suspect this is the culprit. Can you run `accelerate config` and make sure to disable MPS and run the script with `accelerate launch my_script.py`?

**Comment by Naowak** [comment]
I was trying to run the example on MacBook m1, but got the same error. 

For anyone still fighting this issue, I find the solution here : https://discuss.huggingface.co/t/typeerror-cannot-convert-a-mps-tensor-to-float64-dtype-as-the-mps-framework-doesnt-support-float64-please-use-float32-instead/42851.

It worked for me.

**Comment by younesbelkada** [comment]
Thanks @Naowak 
@lvwerra any reason that we use `double()` here: https://github.com/lvwerra/trl/blob/685620ac6cc40159945d690f39706e74f47b5376/trl/trainer/ppo_trainer.py#L1053 ? We should probably change it to float32 as suggested by the linked issue

**Comment by lvwerra** [comment]
Yes, I think that should be fine indeed!

**Comment by younesbelkada** [comment]
Just made: https://github.com/lvwerra/trl/pull/499
